### ISSUE TITLE
Reads from /dev/urandom block some seconds after boot

### ISSUE THREAD

**Issue by nyh** [cloudius-systems/osv#805]
In https://groups.google.com/d/msg/osv-dev/gXt5kbEQTcY/8Y6O_Cp8BAAJ @benoit-canet reports that an application which uses /dev/urandom (a golang application) blocks for many seconds while waiting for the random number generator to be seeded by the host's entropy, which is not yet available.

As @raphaelsc pointed out in that thread, /dev/urandom (unlike /dev/random) is not expected to block while waiting for more high-quality entropy. 

The problem appears to be, though, that when the application wants to use /dev/urandom it thinks there it does not have _any_ seed. Clearly, we can't take a constant seed, or the output of /dev/urandom will be completely predictable, which users do not expect either. It appears that Linux adds early entropy (see also http://stackoverflow.com/questions/33153010/reading-dev-urandom-as-early-as-possible) by adding some not-really-random-but-unique data (like MAC address) to the pool during boot. Linux distributions also save some entropy during shutdown and reread it during reboot (but that doesn't help the first boot).

By why, when the application starts after at least half a second of boot process, do we not have _any_ seed?

My view on this issue is:
1. By the time the application actually starts, a lot of stuff has already happened (OSv startup timings, interrupt times, packet arrivial, dhcp, etc.) that could be used to add entropy to the pool. We don't _have_ to use entropy only from the host (virtio-rng) and can also use these internal sources. We could also have some special boot-time entropy sources especially for solving this issue.
2. Read from /dev/urandom should never block, even if we do not have "enough" entropy yet.

**Comment by tgrabiec** [comment]
@nyh We already have some internal entropy sources - interrupt timing and `rdrand` instruction. I think Benoit's case should be debugged to understand better why we don't get enough entropy for such a long time. 

I don't agree that reads from /dev/urandom should never block. They should block until it is properly seeded, because it is used for cryptographical purposes [1]. 

[1] http://www.2uo.de/myths-about-urandom/

**Comment by nyh** [comment]
I'm not sure what in [1] (which is a long document) proves that our urandom behaves as it should (as it does on Linux).

My point was that I don't know what /dev/urandom should do in very early boot when zero entropy is available (for all I care, it can block like it does now). But by the time boot has finished, roughly 0.5 seconds later, and we started running the application, we should at least have *some* entropy from internal sources. If we don't get any entropy, then I guess we have a bug in our internal sources (?) but if we're getting some, but this some is not "enough", I'm not sure that continuing to block until we have enough is the right thing to do. We're trying to be Linux compatible (not freebsd compatible) so we need to think what Linux does in that case. I think (but not 100% sure) that in that case, Linux's urandom would not block and instead return some "low quality" (but not zero quality) randomness, and if Linux does this, should we too?

**Comment by benoit-canet** [comment]
Go is stuck on that so no activity hence no interrupt.

For the rdrand instruction it's probably defined as a cpu feature and QEMU
is known to pass castrated cpu flags to the guest in order to easy out live
migration.

Would it be correct to modify ./script/run.py to add this feature to the cpu
passwd to the guest ?

**Comment by penberg** [comment]
@nyh if you don't allow `/dev/urandom` to block on lack of entropy, you're basically making it unsafe for generating cryptographic keys, for example, which makes the whole system insecure.

The interesting question here is why you're not getting enough entropy via `virtio-rng`, which taps into _host_ entropy sources.

**Comment by gleb-cloudius** [comment]
On Fri, Oct 28, 2016 at 12:10:13AM -0700, Pekka Enberg wrote:
> @nyh if you don't allow `/dev/urandom` to block on lack of entropy, you're basically making it unsafe for generating cryptographic keys, for example, which makes the whole system insecure.
> 
The same is true for Linux. If you want real entropy use /dev/random.

> The interesting question here is why you're not getting enough entropy via `virtio-rng`, which taps into _host_ entropy sources.
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/cloudius-systems/osv/issues/805#issuecomment-256852026

--
			Gleb.

**Comment by penberg** [comment]
@gleb-cloudius Our `/dev/urandom` doesn't block _after boot_ either. We deliberately picked the FreeBSD behavior over Linux:

http://www.2uo.de/myths-about-urandom/

>Linux's /dev/urandom happily gives you not-so-random numbers before the kernel even had the chance to gather entropy. When is that? At system start, booting the computer.
>
> FreeBSD does the right thing: they don't have the distinction between /dev/random and /dev/urandom, both are the same device. At startup /dev/random blocks once until enough starting entropy has been gathered. Then it won't block ever again.

We did that because OSv is likely to generate cryptographic keys much earlier in boot than Linux because our boot process is so slim. Also, Linux's behavior here is borderline stupid.

**Comment by gleb-cloudius** [comment]
On Sat, Oct 29, 2016 at 11:08:26PM -0700, Pekka Enberg wrote:
> @gleb-cloudius Our `/dev/urandom` doesn't block _after boot_ either. We deliberately picked the FreeBSD behavior over Linux:
> 
I understand the argument, but this breaks Linux compatibility as there
are application that assume that /dev/urandom does not block (they may
well have another code for FreeBSD).

> http://www.2uo.de/myths-about-urandom/
> 
> >Linux's /dev/urandom happily gives you not-so-random numbers before the kernel even had the chance to gather entropy. When is that? At system start, booting the computer.
> >
> > FreeBSD does the right thing: they don't have the distinction between /dev/random and /dev/urandom, both are the same device. At startup /dev/random blocks once until enough starting entropy has been gathered. Then it won't block ever again.
> 
> We did that because OSv is likely to generate cryptographic keys much earlier in boot than Linux because our boot process is so slim. Also, Linux's behavior here is borderline stupid. 
> 
And Nadav says that we should have enough by the time application code
starts running, Linux behaviour is no stupid as long as they have the
means to get real entropy and they do (/dev/random). Also, I think,
you can check how many entropy there is in /dev/urandom right now in
/proc/sys/kernel/random/entropy_avail.

--
			Gleb.

**Comment by nyh** [comment]
Gregor Berginc observed that if we disable virtio-rng in run.py, the hang on golang's boot (while waiting to read from /dev/urandom) disappears. Could it be that even with virtio-rng disabled, our /dev/urandom implementation believes it already got enough entropy during boot from internal sources (interrupt timings, etc.), and yet it nevertheless chooses to **also** wait for more entropy from the host? If that is the case, then it is a bug.

**Comment by nyh** [comment]
I looked at the code. drivers/virtio-rng.cc registers the virtio-rng device as one of the entropy "sources". Its read function, get_random_bytes(), is a blocking function.

The code in bsd/sys/dev/random/live_entropy_sources.cc has a function live_entropy_sources_feed(), which walks through all the sources, reads from each, and feeds whatever we got.  A comment in this functions suggests that "This should be quick, since it's a live entropy source". But it is not - reading from virtio-rng actually can block for a long time. 

live_entropy_sources_feed() is called in random_kthread() in a loop, so there is no reason why it should wait. I think virtio-rng's get_random_bytes() should have just read whatever entropy is available from the host, and not wait for it.

I think the bug we're seeing happening because once random_kthread() hangs waiting for virtio-rng's input, it doesn't continue its loop which is also responsible for processing internal entropy from the ring! So while blocked waiting for virtio-rng, we also ignore the plenty of internal entropy that comes in :-(

I'll try to produce a patch
